{"cells":[{"metadata":{"_uuid":"490f4f04798bf8e91659ab9a9e1dd4d1fa139052","_cell_guid":"70776707-8fec-4f7b-aa73-fb1d96c98b54"},"cell_type":"markdown","source":"# Classifying/Predicting  Molecular Subtypes in Breast Cancer Tumors \n\n"},{"metadata":{"_uuid":"a57c922da3475f028e8e80498f9aaa6a9e803d14","_cell_guid":"4646a8b4-1b5d-4c12-bbae-203b6a9134b1"},"cell_type":"markdown","source":"# Part 3: Loading Packages and Reading in datasets in Rstudio"},{"metadata":{"_kg_hide-output":true,"_uuid":"b60276a0916a107c62defa73f6008dc18e05971e","trusted":true,"_cell_guid":"17f16040-3ce2-4511-9074-957b99d67e09"},"cell_type":"markdown","source":"library(stringr) #assist with text manipulation\nlibrary(dplyr) # data manipulation\nlibrary(readr) # data input\nlibrary(caret) #select tuning parameters\nlibrary(MASS) # contains the data\nlibrary(nnet) # used for Multinomial Classification \nlibrary(readr) #assist with text manipulation\nlibrary(kernlab) #assist with SVM feature selection\nlibrary(class) # used for an object-oriented style of programmin\nlibrary(KernelKnn) # used for K- Nearest-Neighbors method\nlibrary(nnet) # Used for Neural Net\nlibrary(e1071) \nlibrary(gbm)\nlibrary(xgboost) # Used for xgbTree"},{"metadata":{"_kg_hide-output":true,"_uuid":"26244077bfb71f88ec4e439881d33f748e3406c9","trusted":true,"_cell_guid":"ec45a985-a7c3-451b-9b92-acfa0f36cec6"},"cell_type":"code","source":"SSC <-read_csv(\"../input/transposed/77_cancer_proteomes_CPTAC_itraq.csv\")\ndim(SSC) #86 records, 12553 predictors\nclin <-read_csv(\"../input/editpam50/clinical_data_breast_cancer.csv\")\ndim(clin) #105 Records, 30 clinical predictors/measurements","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"257d5feb580e5239184494da046e7faeab09c482","_cell_guid":"417d4c14-1356-40b4-84d1-c011a06f157e"},"cell_type":"markdown","source":"# Part 4: Using Stringr to Manipulate Complete TCGA ID\n \n \n The code below manipulates the Complete TCGA ID number in the Proteomes dataset, queries the Complete TCGA ID number and Molecular tumor type from the Clinical dataset and combines the Proteomes and Clinical dataset as **MASTER**."},{"metadata":{"_uuid":"013dcd9d7da978557e28a60c0c6c8f096e045890","trusted":true,"_cell_guid":"ab1b56a3-6983-40a2-8dcc-b81f01191ec7"},"cell_type":"code","source":"p1<-str_replace(SSC$`Complete TCGA ID`,\"TCGA\",\"\")\np2<-str_replace(p1,\".\\\\d+$\",\"\")\np3<-str_replace(p2,\"^\",\"TCGA-\")\n\nfor (i in seq(1:83)){\n  SSC$`Complete TCGA ID`[i]<-p3[i]\n}\n\ncolnames(clin)\nclinp<-clin[,c(1,21)]\n\nMASTER<- inner_join(SSC,clinp)","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"615fa9d9ad573178a1648665b70aeda0f147c8e5","_cell_guid":"a8939ae8-0ddc-484f-be6d-04a3fa067791"},"cell_type":"markdown","source":"# Part 5: Using Only Variables/Predictors With At Least 99% Of The Data\n\n"},{"metadata":{"_uuid":"ebfb68b349c34fba00a273f7d4c343c20d025df6","trusted":true,"_cell_guid":"80a4d82b-9b21-46b4-b097-ee60c420514b"},"cell_type":"code","source":"#use only variables with 99% of the data\npcent<-sapply(MASTER, function(x) sum(!is.na(x))/length(x))\n              \nclean <- MASTER[1]\nfor (i in seq(2, 12554))\n{\n  if(pcent[i]>0.99){\n    clean<-cbind(clean,MASTER[i])\n  }\n}\n\n# Remove any NA's in the data frame,change to 0\nclean[is.na(clean)] <- 0\nCDF<-clean\n\n#Remove the Complete TCGA ID from the CDF data frame.\n# No longer needer for molecular classification\nCDF<-CDF[-1]\nattach(CDF)","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"776c9519eaa799bde5fb7c7c457c78f30c6f4142","_cell_guid":"adc1b509-0505-4ada-9517-2e775bf42042"},"cell_type":"markdown","source":"\nThe results of the Lasso data reduction technique are too large to be view in the Rstudio console. Therefore the results exported into a txt file on your local machine. "},{"metadata":{"_kg_hide-output":true,"_uuid":"6730c7a9b205825b96e92d488ef72583ef81e830","trusted":true,"_cell_guid":"4c42f220-8e7c-4f16-89eb-5c8aaafb3d0c"},"cell_type":"code","source":"## Randomly dividing our dataset, which encompasses 4 major molecular tumor types, into datasets for testing and training out model. \nlibrary(glmnet)\nCDF$`PAM50 mRNA`<- as.factor(CDF$`PAM50 mRNA`)\nx<-model.matrix(CDF$`PAM50 mRNA`~., CDF)\ntrain<-sample(1:nrow(x), nrow(x)/2)\ntest<-(-train)\ny<-CDF$`PAM50 mRNA`\ny.test<-y[test]\nset.seed(1) \n\n\n#Setting up parameters \ngrid <- 10^seq(10,-2, length =100)\nCDF$`PAM50 mRNA`<- as.factor(CDF$`PAM50 mRNA`)\nset.seed (1)\ncv.out<-cv.glmnet(x[train,],y[train],alpha=1,lambda =grid,family=\"multinomial\",type.multinomial=\"grouped\")\nplot(cv.out)\nbestlam<-cv.out$lambda.min\n\n#  The results of the Lasso data reduction technique are exported onto your local machine for optimal viewing.\nout<-glmnet(x,y,alpha =1, lambda=grid,family=\"multinomial\",type.multinomial=\"grouped\")\nlasso.coef<-predict(out,type =\"coefficients\",s=bestlam,family=\"multinomial\",type.multinomial=\"grouped\")\nsink(file=\"lasso.txt\")\noptions(\"max.print\"=8020)\nlasso.coef\nsink(NULL)","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"e0a42a8a4cb7a5a37a7f889a37e85dbd1226b89c","_cell_guid":"873238e8-e87a-4bbd-a243-803f4975b022"},"cell_type":"markdown","source":"#Reading in and printing the column numbers of all genes for final dataset indexing\nsink(file=\"colindex.txt\")\noptions(\"max.print\"=8100)\ncolnames(CDF)\nsink(NULL)\n\n# List of the 56 genetic predictors and the assoicated column number\n\n`myoferlin isoform a` 142\n`heat shock protein HSP 90-beta isoform a` 258\n`keratin, type II cytoskeletal 72 isoform 1` 283\n`cingulin` 342\n`dedicator of cytokinesis protein 1` 537\n`keratin, type I cytoskeletal 23`774\n`1-phosphatidylinositol 4,5-bisphosphate phosphodiesterase beta-3 isoform 1` 1017\n`TBC1 domain family member 1 isoform 2` 1069\n`kinesin-like protein KIF21A isoform 1` 1107\n`ubiquitin carboxyl-terminal hydrolase 4 isoform a` 1125\n`PREDICTED: myomegalin-like` 1164\n`KN motif and ankyrin repeat domain-containing protein 1 isoform a` 1280\n`spermatogenesis-associated protein 5` 1281\n`TBC1 domain family member 9` 1352\n`receptor tyrosine-protein kinase erbB-2 isoform a precursor` 1376\n`epidermal growth factor receptor isoform a precursor` 1377\n`UPF0505 protein C16orf62` 1572\n`MICAL-like protein 1` 1643\n`UTP--glucose-1-phosphate uridylyltransferase isoform a` 1717\n`probable ATP-dependent RNA helicase DDX6` 1848\n`L-lactate dehydrogenase B chain` 1916\n`myelin expression factor 2` 2037\n`histone-lysine N-methyltransferase NSD3 isoform long` 2065\n`histone-lysine N-methyltransferase NSD3 isoform short` 2067\n`signal transducer and activator of transcription 6 isoform 1` 2540\n`WD repeat-containing protein 91` 2726\n`UPF0553 protein C9orf64` 3201\n`ran-binding protein 9` 3226\n`arfaptin-1 isoform 2` 3438\n`tonsoku-like protein` 3567\n`protein LSM14 homolog B` 3606\n`oxysterol-binding protein-related protein 2 isoform 2` 3629\n`PDZ domain-containing protein GIPC2` 3807\n`calpain small subunit 2` 4075\n`condensin-2 complex subunit G2` 4083\n`telomere length regulation protein TEL2 homolog` 4088\n`guanine nucleotide-binding protein G(olf) subunit alpha isoform 1` 4116\n`squalene synthase` 4613\n`cathepsin B preproprotein` 4816\n`5'-nucleotidase domain-containing protein 2 isoform 1` 4894\n`COBW domain-containing protein 1 isoform 2` 4914\n`transcription elongation factor A protein-like 5`5218\n`DNA repair protein XRCC4 isoform 1` 5542\n`transmembrane protein 132A isoform a precursor` 5574\n`transcription factor AP-2-beta` 5628\n`alpha-methylacyl-CoA racemase isoform 3` 5693\n`trans-acting T-cell-specific transcription factor GATA-3 isoform 1` 6111\n`uncharacterized protein C7orf43` 6179\n`molybdopterin synthase catalytic subunit large subunit MOCS2B` 6533\n`Golli-MBP isoform 1` 6581\n`hepatocyte nuclear factor 3-gamma` 6769\n`39S ribosomal protein L40, mitochondrial`7015\n`migration and invasion enhancer 1`7319\n`ubiquitin-conjugating enzyme E2 E3` 7331\n`protein S100-A13`7357\n`transcription initiation factor TFIID subunit 8`7425\n`phosphoribosyltransferase domain-containing protein 1`7512\n`polyadenylate-binding protein-interacting protein 2`7923"},{"metadata":{"_uuid":"756efd6c71b9bee730d3d6fabc20c04a69192fd2","_cell_guid":"a37e7060-e36d-49ff-8314-5c7dcc6352dc"},"cell_type":"markdown","source":"# Part 8: Selecting Predictors By Column Number \n\n"},{"metadata":{"_uuid":"e173290316f3898a34c468f07b3744fee95d3663","trusted":false,"_cell_guid":"24c56c91-dc74-4496-b301-70e48761eaa0"},"cell_type":"code","source":"testdf<-CDF[,c(8018,142,258,283,537,774,1017,1069,1107,1125,1164,1280,1281,1352,\n               1376,1377,1572,1643,1717,1848,1916,2037,2065,2067,2540,2726,3201,\n               3226,3438,3567,3606,3629,3807,4075,4083,4088,4116,4613,4816,4894,\n               4914,5218,5542,5574,5693,6111,6179,6533,6581,6769,7015,7319,\n               7331,7357,7425,7512,7923)]","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"61078e2fdefe4416a84a01153c3c0f350fb18f04","_cell_guid":"59597a3f-4371-497b-ad9c-fc06881a43af"},"cell_type":"markdown","source":"# Part 9: Classifying Molecular Tumor Type Using Ensemble, Multinomial, K Nearest Neighbors, Support Vector Machine, and Other Machine Learning Methods "},{"metadata":{"_uuid":"4c4e2dddb55c8f07a0a43dddf077c11ffc9bbdbe","_cell_guid":"78925bd0-9dac-42d7-bb5c-6b0fafd301ea"},"cell_type":"markdown","source":"## Molecular Tumor Type Using Ensemble-Boosting Regression Tree \n\nGradient boosting is a machine learning technique for regression and classification problems, the basis involves 3 broad elements. \n\n1. A loss function to be optimized.\n2. A weak learner to make predictions.\n3. An additive model to add weak learners to minimize the loss function.\n\nGradient boosting produces a prediction model in the form of an ensemble of weak prediction models. \n"},{"metadata":{"_uuid":"abc279f7c853aa5bc099f545a2a5febbb637d99d","trusted":true,"_cell_guid":"5936d0f9-ae84-495c-9d32-a0f7e2b0215f"},"cell_type":"code","source":"testdf$`PAM50 mRNA`<- as.factor(testdf$`PAM50 mRNA`)\n\ngbm.model<-gbm(`PAM50 mRNA`~., data=testdf, shrinkage=0.01, distribution = 'multinomial',\n               cv.folds=5, n.trees=5000, verbose=F)\nsummary(gbm.model)\nbest.iter<-gbm.perf(gbm.model, method=\"cv\")","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-output":true,"_uuid":"64c78e89ef8716bd1e8424c8ac2f7d539762d3e0","trusted":true,"_cell_guid":"08f582fb-c6c1-41ae-8d16-534ab254137b"},"cell_type":"code","source":"fitControl <- trainControl(method=\"cv\", number=10, returnResamp = \"all\",classProbs = TRUE, summaryFunction = multiClassSummary)\nAccuraciesGLM<- c(0,0,0)\nfor(i in seq(20)){\n  train<- createDataPartition(testdf$`PAM50 mRNA`, p = .70, list = FALSE)\n  trainDF<-testdf[train,]\n  testDF<-testdf[-train,]\n  trainDF$`PAM50 mRNA`<-as.factor(trainDF$`PAM50 mRNA`)\n  testDF$`PAM50 mRNA`<-as.factor(testDF$`PAM50 mRNA`)\n  gbm.model<-gbm(`PAM50 mRNA`~., data=testdf, shrinkage=0.01, distribution = 'multinomial',\n                 cv.folds=5, n.trees=5000, verbose=F)\n  best.iter<-gbm.perf(gbm.model, method=\"cv\")\n  \n  obj<-train(`PAM50 mRNA`~., data=trainDF, method=\"gbm\",distribution=\"multinomial\",\n             trControl=fitControl, verbose=F,metric = \"ROC\",\n             tuneGrid=data.frame(.n.trees=best.iter, .shrinkage=0.01, .interaction.depth=4, .n.minobsinnode=1))\n  \n  AccuraciesGLM[i] <- confusionMatrix(predict(obj,newdata=testDF[-1]),testDF$`PAM50 mRNA`)$overall[\"Accuracy\"]\n}","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"2513589dcb6021fba07e8a1efa8ed369dd628146","trusted":true,"_cell_guid":"bd544590-2609-44fc-bd9b-fcde38c4b6bd"},"cell_type":"code","source":"plot(density(AccuraciesGLM))\nsummary(AccuraciesGLM)","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"24a8b9de57bb81d892bd5e8a45dbd36ef8265123","_cell_guid":"36dd235f-b726-4057-b20e-f6c1bcf640b8"},"cell_type":"markdown","source":"## Molecular Tumor Type Using K- Nearest-Neighbors \n"},{"metadata":{"_kg_hide-output":true,"_uuid":"b113d08fa6d2dd1d2d1d53238cf9c72a471b25a6","trusted":true,"_cell_guid":"1e5eb211-e558-4a98-b3bc-fa1c49b998be"},"cell_type":"code","source":"library(class)\nlibrary(KernelKnn)\nAccuraciesKNN<-c(0,0,0)\n#choose which measurements to use in classification\ntf<-testdf\nx<-tf[2:57]\n#choose which group labels to use in classification\ny<-factor(testdf$`PAM50 mRNA`)\n\nfor (i in seq(300)){\n  kcres3 <-knn.cv(x, y, k = 3, prob= TRUE)\n  AccuraciesKNN[i] <-confusionMatrix(kcres3,testdf$`PAM50 mRNA`)$overall[\"Accuracy\"]\n}","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"5fcb8161015c3491fffa41f579d6b0fc6c2d009f","trusted":true,"_cell_guid":"d1df57a8-c5c3-4f39-89fe-c23f9417cceb"},"cell_type":"code","source":"summary(AccuraciesKNN)\nplot(density(AccuraciesKNN))","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"ae4177407b90788fdb586b6f8df5e156af44b7c2","_cell_guid":"7ef981ec-a9fa-4ad3-ac85-e6460eefd639"},"cell_type":"markdown","source":"## Molecular Tumor Type Using Support Vector Machines\n\n"},{"metadata":{"_kg_hide-output":true,"_uuid":"f13bc6ebfe233ab7a5f58fcffd17c851b373266d","trusted":true,"_cell_guid":"9969a0b6-2297-47ff-8ea3-d2e9b0d5249e"},"cell_type":"code","source":"library(kernlab)\nAccuraciesSVM <- c(0,0,0)\nfor(i in seq(1000)){\n  \n  train<- createDataPartition(testdf$`PAM50 mRNA`, p = .70, list = FALSE)\n  trainDF<-testdf[train,]\n  testDF<-testdf[-train,]\n  \n  train.ksvm<-ksvm(`PAM50 mRNA` ~.,\n                   scale = TRUE,\n                   data=trainDF,\n                   kernel=\"rbfdot\",\n                   prob.model=TRUE)\n  \n  AccuraciesSVM[i] <- confusionMatrix(testDF$`PAM50 mRNA`, predict(train.ksvm, testDF))$overall[\"Accuracy\"]\n}","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"38c14a3ade1bceda65739336195e256728501ed7","trusted":false,"_cell_guid":"d148f7d5-90a0-4426-9237-a5777b34fd4d"},"cell_type":"code","source":"summary(AccuraciesSVM)\nplot(density(AccuraciesSVM))","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"bc6591aa64e009a07f102fb09db16da89e782904","_cell_guid":"7fa8c0d2-9470-4e71-b36e-103d34ac906e"},"cell_type":"markdown","source":"## Molecular Tumor Type Using Multinomial Method\n\nThe Multinomial method is a classifier that is used to predict the probabilities of the different possible outcomes of a categorically dependent variable. One assumption is that collinearity among independent variables are relatively low."},{"metadata":{"_uuid":"717fa9b9dc09c2b343b9db94eab90a6c77050a33","trusted":false,"_cell_guid":"18e3142d-345b-4dc8-bf69-6726f3d9ca7a"},"cell_type":"code","source":"AccuraciesMultinomial<- c(0,0,0)\nfor(i in seq(1000)){\n  train<- createDataPartition(testdf$`PAM50 mRNA`, p = .70, list = FALSE)\n  trainDF<-testdf[train,]\n  testDF<-testdf[-train,]\n  \n  net<-multinom(`PAM50 mRNA`~.,\n                data=trainDF,trace=FALSE)\n  AccuraciesMultinomial[i] <- confusionMatrix(predict(net, newdata=testDF, \"class\"),testDF$`PAM50 mRNA`)$overall[\"Accuracy\"]\n}","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"1e8bcf837b555342af217b97d0afaca6b143132d","trusted":false,"_cell_guid":"ad39f2ff-bba9-4d02-92ad-46922428225b"},"cell_type":"code","source":"plot(density(AccuraciesMultinomial))\nsummary(AccuraciesMultinomial)","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"b65ec4a2c271eb79de11488031752caa7412887f","_cell_guid":"dd15260b-c91d-4183-b580-e75dc68e4c2f"},"cell_type":"markdown","source":"## Part 10: Conclusions \n\nBelow are the mean results from the machine learning methods,  the models are ran mutlptile times so that one can compute the average the accuracy. \n\n### Support Vector Machines(SVM): 97%\n### Neural Network:  96%\n### Multinomial Method: 90%\n### K- Nearest-Neighbors (KNN): 89%\n\n### eXtreme Gradient Boosting with Random Variables: 55%\n\nThe highest mean classification accuracy resulted from the Support Vector Machines(SVM) and Neural Network model. Models that apply selected predictors from variable selection outperformed the eXtreme Gradient Boosting model with Random Variables in terms of classification accuracy. \n\n"}],"metadata":{"kernelspec":{"display_name":"R","language":"R","name":"ir"},"language_info":{"mimetype":"text/x-r-source","name":"R","pygments_lexer":"r","version":"3.4.2","file_extension":".r","codemirror_mode":"r"}},"nbformat":4,"nbformat_minor":1}